### PR TITLE
Fixing bug with END_STREAM if header has continuations

### LINKED ISSUE #21436: hpack_encoder.cc sets END_STREAM flag on CONTINUATION frame
If a gRPC C++ server's trailers are large enough to be split over a HEADER frame plus subsequent CONTINUATION frames, C++'s implementation sets the END_STREAM flag on the final CONTINUATION frame rather than on the HEADER frame. This is a bug because HTTP/2 CONTINUATION frames do not have an END_STREAM flag (see [relevant section of the spec](https://http2.github.io/http2-spec/#CONTINUATION)); rather, then the END_STREAM should be set on the HEADERS frame itself (the lack of the END_HEADERS flag will let client's know that this really means the stream is over after the subsequent CONTINUATION frames are received). As-is, compliant client implementations will see that the HEADER frame received from a gRPC C++ server does not specify that it is end-of-stream, process the continuations, then dutifully wait (hang forever) until end-of-stream is received on another HEADERS frame that is not going to come.

This issue was reported by a user with a gRPC Java client and a gRPC C++ server ([original issue](https://github.com/grpc/grpc-java/issues/6484)). The reporter's [second comment](https://github.com/grpc/grpc-java/issues/6484#issuecomment-561998631) details the behavior of C++ here, including the lines of code responsible for this behavior.

### PR DESCRIPTION
**Pull request by bocon13**
The HEADER frame should get the END_STREAM flag per the HTTP/2 spec.

fixes #21436

### CODE REVIEW

**Comment by bocon13** [comment]
I've finished all of the expected changes on my side. Can you please rerun the whole test suite so that I can see if there are any other issues to fix?

Thanks!

File: test/core/transport/chttp2/hpack_encoder_test.cc [hunk 1]
```diff
@@ -151,6 +248,53 @@ static void test_basic_headers() {
...
+  grpc_encode_header_options hopt = {0xdeadbeef, /* stream_id */
+                                     is_eof,     /* is_eof */
+                                     false,      /* use_true_binary_metadata */
+                                     150,        /* max_frame_size */
+                                     &stats /* stats */};
+  grpc_chttp2_encode_header(&g_compressor, nullptr, 0, &b, &hopt, &output);
+  verify_frames(output, is_eof);
+  grpc_slice merged;
+  merged = grpc_slice_merge(output.slices, output.count);
+
```
Comment (yashykt) [inline test/core/transport/chttp2/hpack_encoder_test.cc:280]: you shouldn't need this

Comment (bocon13) [reply test/core/transport/chttp2/hpack_encoder_test.cc:280]: good catch. Removed in my next commit

**Comment by bocon13** [comment]
@yashykt no problem and thanks for the review!

Regarding the failing checks, it looks to me to the **ObjC iOS** tests are unrelated timeouts, and the **Interop ALTS** tests are unrelated Docker build errors.

One of the **RBE ASAN C/C++** errors was related, and should be solved by removing the code you requested. I'm not really sure about the other one:
`//test/cpp/qps:json_run_localhost_cpp_generic_async_streaming_qps_unconstrained_1cq_insecure@poller=epollex`

My most recent commit has squashed everything into 1 commit and rebased to latest master.

**Comment by bocon13** [comment]
@yashykt @ericgribkoff thanks for all of the help and reviews!

Can we also backport this to 1.28? I'm new here, and couldn't find the criteria for being included in the support branch. Please let me know if there's anything that you need me to do.

Thanks!

**Comment by yashykt** [comment]
Issues : #18892

**Comment by yashykt** [comment]
Thanks for the fix! @bocon13 if you can open a similar PR for v1.28.x, we can merge it for you

**Comment by bocon13** [comment]
Awesome! Thanks, @yashykt, for merging it!

I've submitted two PRs to backport the fix to the 1.28 and the newly created 1.29 branches.

**Pull request merged**